server/src/main/java/org/elasticsearch/cluster/coordination/JoinReasonService.java

```diff
@@ -0,0 +1,290 @@
+                    }
+                }
+                absentNodes.sort(Comparator.comparing(Tuple::v1));
+                for (int i = discoveryNodes.getSize(); i < absentNodes.size(); i++) {
+                    trackedNodes.remove(absentNodes.get(i).v2());
+                }
+            }
+
+            assert trackedNodes.size() <= discoveryNodes.getSize() * 3;
+        }
+    }
+
+    /**
+     * Called on the master when a {@code node-left} task completes successfully and after the resulting cluster state is applied. If the
+     * absent node is still tracked then this adds the removal reason ({@code disconnected}, {@code lagging}, etc.) to the tracker.
+     */
+    public void onNodeRemoved(DiscoveryNode discoveryNode, String reason) {
+        assert MasterService.isMasterUpdateThread() || Thread.currentThread().getName().startsWith("TEST-")
+            : Thread.currentThread().getName();
+        trackedNodes.computeIfPresent(discoveryNode.getId(), (ignored, trackedNode) -> trackedNode.withRemovalReason(reason));
+    }
+
+    /**
+     * @param discoveryNode The joining node.
+     * @param currentMode   The current mode of the master that the node is joining.
+     * @return              A description of the reason for the join, possibly including some details of its earlier removal.
+     */
+    public String getJoinReason(DiscoveryNode discoveryNode, Coordinator.Mode currentMode) {
+        return trackedNodes.getOrDefault(discoveryNode.getId(), UNKNOWN_NODE)
+            .getDescription(relativeTimeInMillisSupplier.getAsLong(), discoveryNode.getEphemeralId(), currentMode);
+    }
+
+    /**
+     * Get the name of the node and do something reasonable if the node is missing or has no name.
+     */
+    private static String getNodeNameSafe(@Nullable DiscoveryNode discoveryNode) {
+        if (discoveryNode == null) {
+            return "_unknown_";
+        }
+        final String name = discoveryNode.getName();
+        if (Strings.hasText(name)) {
+            return name;
+        }
+        return discoveryNode.getId();
+    }
+
+    private interface TrackedNode {
+
+        TrackedNode present(String ephemeralId);
+
+        TrackedNode absent(long currentTimeMillis, @Nullable DiscoveryNode masterNode);
+
+        TrackedNode withRemovalReason(String removalReason);
+
+        String getDescription(long currentTimeMillis, String joiningNodeEphemeralId, Coordinator.Mode currentMode);
+
+        long getRemovalAgeMillis(long currentTimeMillis);
+    }
+
+    private static final long NOT_REMOVED = -1L;
+
+    private static final TrackedNode UNKNOWN_NODE = new TrackedNode() {
+
+        @Override
+        public TrackedNode present(String ephemeralId) {
+            return new PresentNode(ephemeralId, 0);
+        }
+
+        @Override
+        public TrackedNode absent(long currentTimeMillis, @Nullable DiscoveryNode masterNode) {
+            assert false;
```

Comment:
If we do get this wrong then throwing an exception here will prevent nodes from joining a cluster in production. Since this is just for nicer logging messages I think that's a bit of an extreme reaction, I'd rather just proceed.